### DISCUSSION TITLE
TypeError: Cannot read properties of undefined (reading 'match')

### DISCUSSION THREAD

**Discussion by DavidGarciaCat** [mikro-orm/mikro-orm#3692] [Q&A]
Good day,

I have a TS Service that behaves as a User Retriever using the ORM service to run SQL queries:

```ts
    async findByEmail(email: string): Promise<User> {
        logger.debug(`ORMUserRepository::findByEmail("${email}")`);
        const orm = await ormInit();

        // Adding comment to this line and rewriting the query to find out the reason for the error
        //return orm.em.findOneOrFail(User, { email });

        const user = await orm.em
            .createQueryBuilder(User, 'u')
            .addSelect('u._email').addSelect('u._password').addSelect('u._phpRoles').addSelect('u._roles')
            .where({ 'u._email': email })
            .getSingleResult();
        if (user) return user;
        throw new Error(`User "${email}" not found`);
    }
```

The previous code translates to the following SQL Query (printed on console as I have the `debug` flag set to `true` during the init process):

```sql
select `u`.`email`, `u`.`password`, `u`.`roles`, `u`.`roles_json`
from `users` as `u`
where `u`.`email` = 'mailbox@domain.com'
```

Because my entity looks like this:

```ts
@Entity({ tableName: 'users' })
export default class User extends Aggregate { // The "Aggregate" class provides the "id" property
    @Property({ name: 'email', type: 'string', length: 100, unique: true, nullable: false })
    private _email: string;

    @Property({ name: 'password', type: 'string', length: 255, nullable: false })
    private _password?: string;

    // The "salt" property is part of a legacy database and will need to be deleted as part of an upcoming update.
    // However, it needs to stay here, so I have set this property until the time comes.
    @Property({ name: 'salt', type: 'string', length: 255, nullable: true })
    private _salt: string = 'N/A';

    @Property({ name: 'roles', type: 'string', nullable: true })
    private _phpRoles: string = 'N;';

    @Property({ name: 'roles_json', type: 'json', nullable: true })
    private _roles: UserRole = new UserRole();

    @OneToOne({ entity: () => ReferralCode, mappedBy: 'user', cascade: [ Cascade.PERSIST, Cascade.MERGE ] })
    private readonly _referralCode: ReferralCode = new ReferralCode(this);

    // More properties set below
```

But, after running the SQL query, when the MikroORM files run the `node_modules/@mikro-orm/core/utils/EntityComparator.js` logic, I get the error on the title of this discussion:

```
TypeError: Cannot read properties of undefined (reading 'match')
```

So, I have added some `console.log()` in that file, trying to understand what happened:

```js
// I have replaced the line "node_modules/@mikro-orm/core/utils/EntityComparator.js:201:69":
const propName = (name, parent = 'result') => parent + this.wrap(name);

// By this one:
const propName = (name, parent = 'result') => {
    console.log(`getResultMapper(entityName)` entityName, name);
    parent + this.wrap(name)
};
```

```
getResultMapper(entityName) User id
getResultMapper(entityName) User id
getResultMapper(entityName) User id
getResultMapper(entityName) User email
getResultMapper(entityName) User email
getResultMapper(entityName) User email
getResultMapper(entityName) User password
getResultMapper(entityName) User password
getResultMapper(entityName) User password
getResultMapper(entityName) User salt
getResultMapper(entityName) User salt
getResultMapper(entityName) User salt
getResultMapper(entityName) User roles
getResultMapper(entityName) User roles
getResultMapper(entityName) User roles
getResultMapper(entityName) User roles_json
getResultMapper(entityName) User roles_json
getResultMapper(entityName) User roles_json
getResultMapper(entityName) User undefined
```

And the `Error` thrown exposes this stack trace (before adding the `console.log()` calls, so these are the correct lines):

```
TypeError: Cannot read properties of undefined (reading 'match')
    at EntityComparator.wrap (.../node_modules/@mikro-orm/core/utils/EntityComparator.js:445:17)
    at propName (.../node_modules/@mikro-orm/core/utils/EntityComparator.js:201:69)
    at .../node_modules/@mikro-orm/core/utils/EntityComparator.js:234:44
    at Array.forEach (<anonymous>)
    at EntityComparator.getResultMapper (.../node_modules/@mikro-orm/core/utils/EntityComparator.js:218:20)
    at EntityComparator.mapResult (.../node_modules/@mikro-orm/core/utils/EntityComparator.js:39:29)
    at MariaDbDriver.mapResult (.../node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:47:32)
    at MariaDbDriver.mapResult (.../node_modules/@mikro-orm/knex/AbstractSqlDriver.js:126:27)
    at .../node_modules/@mikro-orm/knex/query/QueryBuilder.js:416:53
    at Array.map (<anonymous>)
```

Given all the properties seem to load in the same order that have been defined, it should now check the `referral code` property, but it targets this `undefined` property/name.

Is there any problem with this Entity configuration? Maybe with my Query?

Thank you very much in advance.

**Comment by B4nan** [comment]
This is an inverse side of relation, how does the owning side look like?

```ts
  @OneToOne({ entity: () => ReferralCode, mappedBy: 'user', cascade: [ Cascade.PERSIST, Cascade.MERGE ] })
  private readonly _referralCode: ReferralCode = new ReferralCode(this);
```

Please prepare complete but minimal repro, could be a bug, sounds similar to another reported issue with overriding the column names (as I see you do it pretty much on every property).

**Reply by DavidGarciaCat** [reply]
The referral code extends another abstract class:

```ts
export default abstract class AbstractCode extends Aggregate {
    private static readonly CODE_LENGTH: number = 8;
    private static readonly HASH_LENGTH: number = 20;

    @Property({ name: 'short_code', type: 'string', length: AbstractCode.CODE_LENGTH, unique: true, nullable: false })
    private _code: string = '';

    @Property({ name: 'hash_code', type: 'string', length: AbstractCode.HASH_LENGTH, unique: true, nullable: false })
    private _hash: string = '';

    @Property({ name: 'active', type: 'boolean', nullable: false })
    private _active: boolean = true;

    @Property({ name: 'created_at', type: 'datetime', nullable: false })
    private readonly _createdAt: DateTime = currentDateTime();
```

```ts
export default class ReferralCode extends AbstractCode {
    @OneToOne({ entity: () => User, inversedBy: 'referralCode', referenceColumnName: 'id', name: 'user_id', unique: true })
    public readonly user: User;
```

I have tried with both implementations, but both seem to fail:

```ts
    private readonly _referralCode: ReferralCode = new ReferralCode(this);

    get referralCode(): ShortCodeReferral {
        return this._referralCode;
    }
```

```ts
    public readonly referralCode: ReferralCode = new ReferralCode(this);
```

**Reply by DavidGarciaCat** [reply]
The reason for overriding the column names is, essentially, that I need to deploy this NodeJS App using an existing database (which is not open to debate if it can be migrated to a new database/table, unfortunately, due to the volume of data), so I need to make sure all columns are mapped properly and not causing any error due to the nomenclature and normalization.

**Reply by B4nan** [reply]
> inversedBy: 'referralCode'

Here you map it to `referralCode` but your property on the other side is called `_referralCode`.

**Reply by B4nan** [reply]
It can't be mapped to some getter, it needs to be mapped to the property that is decorated, which is `_referralCode`.

**Reply by B4nan** [reply]
If you want more help, I need to see a complete but minimal reproduction. What you gave me is very complex as well as very incomplete.

**Reply by DavidGarciaCat** [reply]
I also considered if target the getter could be the reason, and that's why I mentioned I tried both:

> <img width="855" alt="Screenshot 2022-11-04 at 17 57 30" src="https://user-images.githubusercontent.com/7809429/200032670-4e9741ef-a1c8-4b53-a9a3-bc6009fd1794.png">

After making the following changes:
- User
  - remove the getter
  - make the property public
  - rename the property without underscore

I'm still seeing this error message:

```
  ==> Entity Name: "User" ==> Name: "roles_json"
  ==> Entity Name: "User" ==> Name: "undefined"
TypeError: Cannot read properties of undefined (reading 'match')
```

I'll try to create a micro-project with a similar environment, but it might take a bit of time.

**Reply by DavidGarciaCat** [reply]
Good day,

A bit of a follow up for this discussion we started:

After removing the inversed ORM decorator (so, I **cannot** run `user.referralCode`) I have got another error:

```
TypeError: Cannot read properties of undefined (reading 'compositePK')
    at EntityFactory.findEntity (.../node_modules/@mikro-orm/core/entity/EntityFactory.js:182:19)
    at EntityFactory.create (.../node_modules/@mikro-orm/core/entity/EntityFactory.js:37:29)
    at SqlEntityManager.merge (.../node_modules/@mikro-orm/core/EntityManager.js:495:73)
    at SqlEntityManager.map (.../node_modules/@mikro-orm/core/EntityManager.js:476:21)
    at .../node_modules/@mikro-orm/knex/query/QueryBuilder.js:439:37
    at Array.map (<anonymous>)
    at QueryBuilder.getResultList (.../node_modules/@mikro-orm/knex/query/QueryBuilder.js:439:20)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async command (.../src/command/open-search/reindex-content.ts:22:25)
```

Is there any suggestion for this case that can bring further details (or just solve the problem)?

The query I am trying to run is quite simple:

```ts
const res = await orm.em.createQueryBuilder(User, 'u').limit(2).getResult();
```

Thanks in advance for any help you can provide.

==== UPDATED ====

it seems to fail when running the conditional, because `meta` is undefined:

```ts
    findEntity(data, meta, options) {
        const schema = this.driver.getSchemaName(meta, options);
        console.log(meta); // `meta` returns undefined
        if (!meta.compositePK && !meta.getPrimaryProps()[0]?.customType) {
```

**Reply by B4nan** [reply]
Let me say it one more time:

If you want more help, I need to see a complete but minimal reproduction.

**Reply by DavidGarciaCat** [reply]
This leads to a problem, because when I create a small project with the same mapping it was causing an error, it seems to work, as you can see here: https://github.com/DavidGarciaCat/mikro-orm-example-mapped-entities

The output I get is as follows:
```
npm start

> start
> nodemon src/app.ts

[nodemon] 2.0.20
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: ts,json
[nodemon] starting `ts-node src/app.ts`
Data has been created
==== LOADED DATA ==== [
  User {
    uuid: '018629f9-dbba-4561-a0a2-664bd977904b',
    createdAt: '2022-11-08 13:40:21',
    username: 'dd217224-5d1f-49b2-b8aa-3d8eb554a694',
    password: '4cc36775-c572-4632-8291-44da2d623796',
    referralCode: (ReferralCode) { uuid: 'c7911903-3205-4426-b92e-540f7b302d0f', user: [User] }
  },
  User {
    uuid: 'd028ad91-f805-4fc4-9f0a-002d576f4043',
    createdAt: '2022-11-08 13:40:21',
    username: '00b49365-ccfe-42a1-bf7f-c5b40dff1580',
    password: '1568db6a-5444-48a9-885e-c20743dab638',
    referralCode: (ReferralCode) { uuid: 'cdc60161-d545-4c9c-86aa-76762e52e882', user: [User] }
  },
  User {
    uuid: 'f9e93594-f593-4d6c-80cc-a53b776fdd15',
    createdAt: '2022-11-08 13:40:21',
    username: 'f600f76d-9a79-44d7-bf26-32184243c91a',
    password: 'f01ffa5d-cf14-4828-a0db-e923ab699065',
    referralCode: (ReferralCode) { uuid: '1fead463-17cb-438e-83c4-58497f622173', user: [User] }
  }
]
```

However, I get the already mentioned error when I try to use the same approach on another project, that has an existing database. I am almost sure there must be an error mapping one or more entities (although I have reviewed them and I am 99% sure I didn't miss any `public`/`private` property with their mapping).

If you have an idea on how I can show you the error without exposing a private repository for a business app, I'm happy to consider it.

**Reply by B4nan** [reply]
> This leads to a problem, because when I create a small project with the same mapping it was causing an error, it seems to work, as you can see here: https://github.com/DavidGarciaCat/mikro-orm-example-mapped-entities

That's exactly why I need you need to provide failing reproduction :] 

To check if your entity definition is matching existing schema, try running `npx mikro-orm schema:update -d`, if it prints something, the schema is out of sync.

> If you have an idea on how I can show you the error without exposing a private repository for a business app, I'm happy to consider it.

I am not interested in debugging your app, not as part of free support. You have a failing app, clone it, start simplifying as much as you can, and you will end up with a repro, or you find the problem in your code on the way.

**Reply by DavidGarciaCat** [reply]
> I am not interested in debugging your app, not as part of free support. You have a failing app, clone it, start simplifying as much as you can, and you will end up with a repro, or you find the problem in your code on the way.

And I was not suggesting it either; that's why I was not sharing it from the beginning :-)

> To check if your entity definition is matching existing schema, try running `npx mikro-orm schema:update -d`, if it prints something, the schema is out of sync.

I can see a `SqlError`, so I will try to debug it for now and let's see what I can find. Thanks for pointing to the `schema:update` command!
